File: examples/ios-swift/OpenIMLifecycleExample.swift
import Foundation

enum OpenIMLifecycleError: Error {
    case createFailed
    case nativeError(Int32, String)
}

private func openIMLifecycleEventCallback(
    _ userData: UnsafeMutableRawPointer?,
    _ event: UnsafePointer<CChar>?,
    _ payloadJSON: UnsafePointer<CChar>?
) {
    _ = userData
    _ = event.map { String(cString: $0) }
    _ = payloadJSON.map { String(cString: $0) }
}

final class OpenIMLifecycleExample {
    private var handle: OpaquePointer?

    deinit {
        if let handle {
            openim_session_destroy(handle)
        }
    }

    func run(apiAddr: String, wsAddr: String, userID: String, token: String) throws {
        let session = try apiAddr.withCString { api in
            try wsAddr.withCString { ws in
                guard let session = openim_session_create(api, ws, OPENIM_PLATFORM_IOS) else {
                    throw OpenIMLifecycleError.createFailed
                }
                return session
            }
        }
        handle = session

        let listenerID = openim_session_register_listener(session, openIMLifecycleEventCallback, nil)
        guard listenerID != 0 else {
            throw OpenIMLifecycleError.nativeError(
                OPENIM_FFI_ERROR,
                openim_session_last_error(session).map { String(cString: $0) } ?? ""
            )
        }

        try check(openim_session_init(session), session)
        try userID.withCString { user in
            try token.withCString { authToken in
                try check(openim_session_login(session, user, authToken), session)
            }
        }
        try check(openim_session_logout(session), session)
        try check(openim_session_uninit(session), session)
        try check(openim_session_unregister_listener(session, listenerID), session)

        openim_session_destroy(session)
        handle = nil
    }

    private func check(_ code: Int32, _ session: OpaquePointer) throws {
        guard code == OPENIM_FFI_OK else {
            let message = openim_session_last_error(session).map { String(cString: $0) } ?? ""
            throw OpenIMLifecycleError.nativeError(code, message)
        }
    }
}

File: examples/android-kotlin/OpenIMLifecycleExample.kt
package io.openim.example

class OpenIMLifecycleExample(
    private val bridge: OpenIMNativeBridge = OpenIMNativeBridge()
) {
    fun run(apiAddr: String, wsAddr: String, userID: String, token: String) {
        val handle = bridge.openimSessionCreate(apiAddr, wsAddr, OpenIMNativeBridge.PLATFORM_ANDROID)
        require(handle != 0L) { "OpenIM session create failed" }

        try {
            bridge.check(bridge.openimSessionInit(handle), handle)
            bridge.check(bridge.openimSessionLogin(handle, userID, token), handle)
            bridge.check(bridge.openimSessionLogout(handle), handle)
            bridge.check(bridge.openimSessionUninit(handle), handle)
        } finally {
            bridge.openimSessionDestroy(handle)
        }
    }
}

class OpenIMNativeBridge {
    external fun openimSessionCreate(apiAddr: String, wsAddr: String, platformID: Int): Long
    external fun openimSessionDestroy(handle: Long)
    external fun openimSessionInit(handle: Long): Int
    external fun openimSessionLogin(handle: Long, userID: String, token: String): Int
    external fun openimSessionLogout(handle: Long): Int
    external fun openimSessionUninit(handle: Long): Int
    external fun openimSessionLastError(handle: Long): String

    fun check(code: Int, handle: Long) {
        require(code == OK) { openimSessionLastError(handle) }
    }

    companion object {
        const val OK = 0
        const val PLATFORM_ANDROID = 2

        init {
            System.loadLibrary("openim_android_example")
        }
    }
}
